src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2342 @@
+    /// neighbor, then also update its in/out-degree estimates in the peers DB.
+    /// Returns Ok(true) if we're done
+    /// Returns Ok(false) if we're still waiting
+    /// Returns Err(..) on irrecoverable error
+    pub fn getneighbors_neighbors_try_finish(
+        &mut self,
+        network: &mut PeerNetwork,
+    ) -> Result<bool, net_error> {
+        assert!(self.state == NeighborWalkState::GetNeighborsNeighborsFinish);
+
+        // see if we got any replies
+        for (naddr, message) in self.collect_replies(network).into_iter() {
+            let nkey = naddr.to_neighbor_key(network);
+            match message.payload {
+                StacksMessageType::Neighbors(ref data) => {
+                    debug!(
+                        "{:?}: Got Neighbors from {:?}: {:?}",
+                        network.get_local_peer(),
+                        &nkey,
+                        &data.neighbors
+                    );
+                    let neighbors = NeighborWalk::filter_sensible_neighbors(data.neighbors.clone());
+                    self.resolved_getneighbors_neighbors
+                        .insert(naddr, neighbors);
+                }
+                StacksMessageType::Nack(ref data) => {
+                    // not broken; likely because it hasn't gotten to processing our
+                    // handshake yet.  We'll just ignore it.
+                    debug!(
+                        "{:?}: Neighbor {:?} NACKed with code {:?}",
+                        network.get_local_peer(),
+                        &nkey,
+                        data.error_code
+                    );
+                }
+                _ => {
+                    // unexpected reply
+                    debug!("{:?}: Neighbor {:?} replied an out-of-sequence message (type {}); assuming broken", network.get_local_peer(), &nkey, message.get_message_name());
+                    self.add_broken(network, &nkey);
+                }
+            }
+        }
+
+        if self.count_inflight() > 0 {
+            // not done yet
+            debug!(
+                "{:?}: still waiting for {} Neighbors replies",
+                network.get_local_peer(),
+                self.count_inflight()
+            );
+            return Ok(false);
+        }
+
+        // finished!  build up frontier's in-degree estimation, plus ourselves
+        self.cur_neighbor.in_degree = 1;
+        self.cur_neighbor.out_degree = self.frontier.len() as u32;
+
+        for (_, neighbor_list) in self.resolved_getneighbors_neighbors.iter() {
+            for na in neighbor_list {
+                if na.addrbytes == self.cur_neighbor.addr.addrbytes
+                    && na.port == self.cur_neighbor.addr.port
+                {
+                    self.cur_neighbor.in_degree += 1;
+                }
+            }
+        }
+
+        // only save if the neighbor is routable from us
+        if self.walk_outbound {
```

Comment:
Actually, I'm inclined to remove this check.  We should save all discovered neighbors, because we can still walk to them even if they're not routable (i.e. we'd do so via an inbound neighbor walk).

Reply:
On second thought, we don't want to save a neighbor if (1) it's the first neighbor the walk talks to and (2) the neighbor is inbound.  This is because the `frontier` table contents are served to other neighbors in response to a `GetNeighbors` request.  It would do us no good if the `frontier` table got populated with unroutable entries.  To avoid this, we'd avoid saving `cur_neighbor` if it might be unroutable (which only occurs if (1) and (2) are both true).  However, after stepping from `cur_neighbor` to a subsequent neighbor, the subsequent neighbor is definitely routable from us because we would have already handshaked with it (so we can and should save it).